dkc/core/models/quota.py

```diff
@@ -0,0 +1,64 @@
+from typing import Type
+
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.core.exceptions import ValidationError
+from django.db import IntegrityError, models, transaction
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+
+def _default_user_quota() -> int:
+    return settings.DKC_DEFAULT_QUOTA
+
+
+class Quota(models.Model):
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                check=models.Q(used__lte=models.F('allowed')),
+                name='used_lte_allowed',
+            )
+        ]
+
+    used = models.PositiveBigIntegerField(default=0)
+    allowed = models.PositiveBigIntegerField(default=_default_user_quota)
+
+    # Nullable OneToOneField ensures that only one Quota per User exists
+    user = models.OneToOneField(
```

Comment:
I thought the intent here was that the quota relation should be decoupled from users and folders, and that users and folders could instead point to the quota table? Having a user column here violates that design.

Reply:
I sort of agree in principle. The practical difficulty is that using a custom Django `User` model is quite burdensome. The better approach is to create a `Profile` model, point it at a `User` with a `OneToOneField` relation, then tack your application-specific fields onto the `Profile`. However, even adding a new `Profile` model adds some amount of engineering burden (particularly around its automatic creation, usage with test factories, etc.), so I took this simpler approach for now. I'm sure we'll eventually need a `Profile`, but it'd be easier to add incrementally.

One ostensible weakness here is that we don't have a database-level guarantee that every `User` has a `Quota`. However, even with a `Profile,` we actually don't have that guarantee either (we can guarantee that every `Profile` has a `Quota`, but not that every `User` has a `Profile` and hence a `Quota`).

One surprising advantage of this method is that it's very easy to filter `Quota` by which have a null `user`. This allows us to distinguish "project quotas" from "user quotas", which might be useful administratively, and which is not practical if the relation goes in the reverse direction. Once I saw is, I actually realize that semantically, it does kind of make sense for `Quota` to explicitly "know about" `User`, since we often talk about "user quotas" in our design discussions.

Reply:
> One surprising advantage of this method is that it's very easy to filter Quota by which have a null user. This allows us to distinguish "project quotas" from "user quotas"

Such semantics for null value strikes me as poor database design -- some would even call it an anti-pattern in RDBMS. The better way to distinguish project quotas from user quotas would be to do so explicitly, rather than trying to force it using a null value. I realize this stems from an idiosyncrasy of Django, but I think it would be unwise for us to let such an idiosyncrasy drive our database schema in a direction we don't really want it to go.